dora/underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUfsStatusIterator.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.NoSuchElementException;
+
+/**
+ * HDFS under file system status iterator.
+ */
+public class HdfsUfsStatusIterator implements Iterator<UfsStatus> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HdfsUfsStatusIterator.class);
+
+  private final FileSystem mFs;
+
+  /**
+   * Each element is a pair of (full path, UfsStatus).
+   */
+  private LinkedList<Pair<String, UfsStatus>> mDirPathsToProcess = new LinkedList<>();
+
+  private RemoteIterator<FileStatus> mHdfsRemoteIterator;
+
+  private final long mUserBlockSizeBytesDefault =
+      Configuration.global().getBytes(PropertyKey.USER_BLOCK_SIZE_BYTES_DEFAULT);
+
+  /**
+   * HDFS under file system status iterator.
+   * @param path the path for listing
+   * @param fs the hdfs file system
+   */
+  public HdfsUfsStatusIterator(String path, FileSystem fs) {
+    mFs = fs;
+    initQueue(path);
+  }
+
+  private void initQueue(String path) {
+    try {
+      Path thePath = new Path(path);
+      mHdfsRemoteIterator = mFs.listStatusIterator(thePath);
+    } catch (IOException e) {
+      LOG.error("Failed to list the path {}", path, e);
+    }
+  }
+
+  @Override
+  public boolean hasNext() {
+    try {
+      if (mHdfsRemoteIterator.hasNext()) {
+        return true;
+      }
+      if (mDirPathsToProcess.isEmpty()) {
+        return false;
+      }
+      while (!mDirPathsToProcess.isEmpty()) {
+        Pair<String, UfsStatus> dir = mDirPathsToProcess.removeFirst();
+        String path = dir.getFirst();
+        mHdfsRemoteIterator = mFs.listStatusIterator(new Path(path));
+        if (mHdfsRemoteIterator.hasNext()) {
+          return true;
+        }
+      }
+      return false;
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public UfsStatus next() {
+    if (!hasNext()) {
+      throw new NoSuchElementException();
+    }
+    try {
+      FileStatus fileStatus = mHdfsRemoteIterator.next();
+      UfsStatus ufsStatus;
+      Path path = fileStatus.getPath();
+      AlluxioURI alluxioUri = new AlluxioURI(path.toString());
+      if (fileStatus.isDirectory()) {
+        ufsStatus = new UfsDirectoryStatus(path.getName(), fileStatus.getOwner(),
+            fileStatus.getGroup(), fileStatus.getPermission().toShort(),
+            fileStatus.getModificationTime());
+        mDirPathsToProcess.addLast(new Pair<>(path.toString(), ufsStatus));
+      } else {
+        ufsStatus = new UfsFileStatus(path.getName(), alluxioUri.hash(), fileStatus.getLen(),
+            fileStatus.getModificationTime(), fileStatus.getOwner(), fileStatus.getGroup(),
+            fileStatus.getPermission().toShort(), mUserBlockSizeBytesDefault);
```

Comment:
shouldn't be block size be the HDFS block size instead alluxio one?

Reply:
+1

Reply:
Updated. I used HDFS block size instead.